src/nvme/ioctl.c

```diff
@@ -594,48 +594,61 @@ int nvme_nvm_identify_ctrl(int fd, struct nvme_id_ctrl_nvm *id)
 	return nvme_identify_ctrl_csi(fd, NVME_CSI_NVM, id);
 }
 
-int nvme_get_log(int fd, enum nvme_cmd_get_log_lid lid, __u32 nsid, __u64 lpo,
-		 __u8 lsp, __u16 lsi, bool rae, __u8 uuidx, enum nvme_csi csi,
-		 bool ot, __u32 len, void *log, __u32 timeout, __u32 *result)
+int nvme_get_log(struct nvme_get_log_args *args)
 {
-	__u32 numd = (len >> 2) - 1;
+	__u32 numd = (args->len >> 2) - 1;
 	__u16 numdu = numd >> 16, numdl = numd & 0xffff;
 
-	__u32 cdw10 = NVME_SET(lid, LOG_CDW10_LID) |
-			NVME_SET(lsp, LOG_CDW10_LSP) |
-			NVME_SET(!!rae, LOG_CDW10_RAE) |
+	__u32 cdw10 = NVME_SET(args->lid, LOG_CDW10_LID) |
+			NVME_SET(args->lsp, LOG_CDW10_LSP) |
+			NVME_SET(!!args->rae, LOG_CDW10_RAE) |
 			NVME_SET(numdl, LOG_CDW10_NUMDL);
 	__u32 cdw11 = NVME_SET(numdu, LOG_CDW11_NUMDU) |
-			NVME_SET(lsi, LOG_CDW11_LSI);
-	__u32 cdw12 = lpo & 0xffffffff;
-	__u32 cdw13 = lpo >> 32;
-	__u32 cdw14 = NVME_SET(uuidx, LOG_CDW14_UUID) |
-			NVME_SET(csi, LOG_CDW14_CSI) |
-			NVME_SET(!!ot, LOG_CDW14_OT);
+			NVME_SET(args->lsi, LOG_CDW11_LSI);
+	__u32 cdw12 = args->lpo & 0xffffffff;
+	__u32 cdw13 = args->lpo >> 32;
+	__u32 cdw14 = NVME_SET(args->uuidx, LOG_CDW14_UUID) |
+			NVME_SET(!!args->ot, LOG_CDW14_OT) |
+			NVME_SET(args->csi, LOG_CDW14_CSI);
 
 	struct nvme_passthru_cmd cmd = {
 		.opcode		= nvme_admin_get_log_page,
-		.nsid		= nsid,
-		.addr		= (__u64)(uintptr_t)log,
-		.data_len	= len,
+		.nsid		= args->nsid,
+		.addr		= (__u64)(uintptr_t)args->log,
+		.data_len	= args->len,
 		.cdw10		= cdw10,
 		.cdw11		= cdw11,
 		.cdw12		= cdw12,
 		.cdw13		= cdw13,
 		.cdw14		= cdw14,
-		.timeout_ms	= timeout,
+		.timeout_ms	= args->timeout,
 	};
 
-	return nvme_submit_admin_passthru(fd, &cmd, result);
+	if (args->args_size < sizeof(struct nvme_get_log_args))
+		return -EINVAL;
+	return nvme_submit_admin_passthru(args->fd, &cmd, args->result);
 }
 
 static int __nvme_get_log(int fd, enum nvme_cmd_get_log_lid lid, bool rae,
 			  __u32 len, void *log)
 {
-	return nvme_get_log(fd, lid, NVME_NSID_ALL, 0, NVME_LOG_LSP_NONE,
-			    NVME_LOG_LSI_NONE, rae, NVME_UUID_NONE,
-			    NVME_CSI_NVM, false, len, log,
-			    NVME_DEFAULT_IOCTL_TIMEOUT, NULL);
+	struct nvme_get_log_args args = {
+		.fd = fd,
+		.lid = lid,
+		.nsid = NVME_NSID_ALL,
+		.lpo = 0,
+		.lsp = NVME_LOG_LSP_NONE,
+		.lsi = NVME_LOG_LSI_NONE,
+		.rae = rae,
+		.uuidx = NVME_UUID_NONE,
+		.len = len,
+		.log = log,
+		.timeout = NVME_DEFAULT_IOCTL_TIMEOUT,
+		.result = NULL,
+		.csi = NVME_CSI_NVM,
+		.ot = false,
```

Comment:
I think I would move the timeout and result to the end of the struct to make it more consistent with the rest.

(Since I am already working on it I'll do it)

Reply:
Thinking on it a bit more, maybe it would make more sense to have the common ones at the beginning of the struct. This would be 'args_size', fd', 'timeout and 'result'.